odoo/tools/misc.py

```diff
@@ -135,124 +135,66 @@ def exec_pg_command_pipe(name, *args):
-    @param mode file open mode
-    @param subdir subdirectory
-    @param pathinfo if True returns tuple (fileobject, filepath)
-
-    @return fileobject if pathinfo is False else (fileobject, filepath)
+    :param str file_path: absolute file path, or relative path within any `addons_path` directory
+    :param list(str) filter_ext: optional list of supported extensions (without leading dot)
+    :return: the absolute path to the file
+    :raise IOError: if the file is not found under the known `addons_path` directories
+    :raise ValueError: if the file doesn't have one of the supported extensions (`filter_ext`)
     """
-    adps = odoo.addons.__path__
-    rtp = os.path.normcase(os.path.abspath(config['root_path']))
-
-    basename = name
-
-    if os.path.isabs(name):
-        # It is an absolute path
-        # Is it below 'addons_path' or 'root_path'?
-        name = os.path.normcase(os.path.normpath(name))
-        for root in adps + [rtp]:
-            root = os.path.normcase(os.path.normpath(root)) + os.sep
-            if name.startswith(root):
-                base = root.rstrip(os.sep)
-                name = name[len(base) + 1:]
-                break
-        else:
-            # It is outside the OpenERP root: skip zipfile lookup.
-            base, name = os.path.split(name)
-        return _fileopen(name, mode=mode, basedir=base, pathinfo=pathinfo, basename=basename)
-
-    if name.replace(os.sep, '/').startswith('addons/'):
-        subdir = 'addons'
-        name2 = name[7:]
-    elif subdir:
-        name = os.path.join(subdir, name)
-        if name.replace(os.sep, '/').startswith('addons/'):
-            subdir = 'addons'
-            name2 = name[7:]
+    root_path = os.path.abspath(config['root_path'])
+    addons_path = odoo.addons.__path__ + [root_path]
+    abs_path = os.path.isabs(file_path)
+
+    # ignore leading 'addons/' if present, it's the final component of root_path, but
+    # may sometimes be included in relative paths
+    if file_path.startswith('addons' + os.sep):
+        file_path = file_path[7:]
+
+    for root in addons_path:
+        # final path sep required to avoid partial match
+        root_path = os.path.normpath(os.path.normcase(root)) + os.sep
+
+        if abs_path:
+            fpath = os.path.normpath(os.path.normcase(file_path))
         else:
-            name2 = name
+            fpath = os.path.normpath(os.path.normcase(os.path.join(root, file_path)))
 
-    # First, try to locate in addons_path
-    if subdir:
-        for adp in adps:
-            try:
-                return _fileopen(name2, mode=mode, basedir=adp,
-                                 pathinfo=pathinfo, basename=basename)
-            except IOError:
-                pass
+        if fpath.startswith(root_path) and os.path.exists(fpath):
+            if filter_ext and fpath.lower().rsplit(".")[-1] not in filter_ext:
+                raise ValueError("Unsupported path: " + file_path)
+            return fpath
```

Comment:
Same suggestion as elsewhere: why not require the leading `.` in the extensions, make the default `filter_ext=('',)` and then just use `fpath.lower().endswith(filter_ext)`?

Also maybe run the filter first, that way we only pay for the `stat` call if the path is suitable? Plus "Unsupported path" might make more sense if the path doesn't match `root_path`.

Reply:
I did that because it feels redundant to ask callers to specify the leading dot in the extensions list, but either works for me

Reply:
> Also maybe run the filter first, that way we only pay for the `stat` call if the path is suitable? Plus "Unsupported path" might make more sense if the path doesn't match `root_path`.

I did want a `ValueError` when the file exists but doesn't have an allowed extension. But we can probably short-circuit that and raise the `ValueError` immediately at the start, regardless of whether the file exists or not. Sounds just as consistent, and slightly cheaper indeed :)